File: pkg/client-go/meshbench/sim.go
// The clock, and the run.
package meshbench

import (
	"context"
	"fmt"
	"time"
)

// Sim is the run. Live.
type Sim struct{ w *Workbench }

// Sim reaches the clock.
func (w *Workbench) Sim() Sim { return Sim{w} }

// State is what the clock is doing.
func (s Sim) State(ctx context.Context) (SimState, error) {
	var st SimState
	return st, s.w.CallInto(ctx, "sim.state", nil, &st)
}

// Start brings the run up: waits out the warm, starts every node, and plays.
//
// Deliberately not one call to sim.start. That verb is the play button's own
// handler and answers four ways - it pauses if already playing, declines while
// links are being measured, or starts firmware and does not play - so a script
// pressing it once gets whichever of those the moment happens to be in.
//
// Worse, it only starts firmware when *no* node is running. Pin a build onto
// two nodes of a fifty-eight node fixture and it considers the mesh started,
// plays with fifty-six of them down, and says nothing.
//
// So this asks for the three things it actually wants, in order, and checks
// each one. Zero timeouts mean the usual ones.
func (s Sim) Start(ctx context.Context) error {
	return s.StartWithin(ctx, 0, 0)
}

// StartWithin is Start with the two waits named: how long to give the link
// measurement, and how long to give the firmware.
func (s Sim) StartWithin(ctx context.Context, warm, firmware time.Duration) error {
	if warm <= 0 {
		warm = 30 * time.Minute
	}
	// The links first. Nothing that follows means anything against a matrix
	// that is still being measured.
	if err := s.w.WaitIdle(ctx, warm); err != nil {
		return err
	}
	// Idle is not the same as measured. A warm that failed or was cancelled
	// finishes its own job row, so the wait above returns having waited for
	// nothing: no link was measured, and every study after this would answer
	// over free space.
	if st, err := s.State(ctx); err != nil {
		return err
	} else if !st.LinksMeasured && !st.Warming {
		note := st.Ground.Note
		if note == "" {
			note = "warm the links again before reading anything from this run"
		}
		return fmt.Errorf("no link has been measured, so nothing here can "+
			"reach anything. %s", note)
	}
	// Then every node that is not up, which firmware.start does and sim.start
	// does only when none of them are.
	st, err := s.w.Firmware().State(ctx)
	if err != nil {
		return err
	}
	if st.Running < st.Nodes {
		if err := s.w.Firmware().Start(ctx); err != nil {
			return err
		}
		if err := s.w.Firmware().WaitStarted(ctx, firmware); err != nil {
			return err
		}
	}
	// Then the clock, by its own name. Play cannot pause, which is the other
	// half of what made Start unusable from a script.
	now, err := s.State(ctx)
	if err != nil {
		return err
	}
	if now.Playing {
		return nil
	}
	return s.Play(ctx)
}

// Play, Pause and Toggle are the clock itself.
func (s Sim) Play(ctx context.Context) error   { return s.w.Do(ctx, "sim.play", nil) }
func (s Sim) Pause(ctx context.Context) error  { return s.w.Do(ctx, "sim.pause", nil) }
func (s Sim) Toggle(ctx context.Context) error { return s.w.Do(ctx, "sim.toggle", nil) }

// Step advances one tick, which is StepMs of simulated time.
func (s Sim) Step(ctx context.Context) error { return s.w.Do(ctx, "sim.step", nil) }

// Reset puts the clock and the counters back to the start of the run.
func (s Sim) Reset(ctx context.Context) error { return s.w.Do(ctx, "sim.reset", nil) }

// SetSeed fixes the run. Same seed, same scenario, same result - which is what
// makes a *changed* result mean something.
func (s Sim) SetSeed(ctx context.Context, seed uint64) error {
	return s.w.Do(ctx, "sim.seed", map[string]any{"seed": float64(seed)})
}

// SetStepMs is how much simulated time one tick advances.
func (s Sim) SetStepMs(ctx context.Context, ms uint32) error {
	return s.w.Do(ctx, "sim.speed", map[string]any{"step_ms": float64(ms)})
}

// SetRealFirmware chooses whether play starts MeshCore on every node, or runs
// the channel with nothing behind it.
func (s Sim) SetRealFirmware(ctx context.Context, on bool) error {
	return s.w.Do(ctx, "sim.kind", map[string]any{"real": on})
}

// Run advances the mesh's own clock by this much and waits for it to finish.
//
// Simulated time, not yours. Five minutes here is five minutes of the mesh's
// clock; on 155 emulated nodes that is a great deal longer than five of yours,
// which is why the wait's own timeout is separate and generous.
func (s Sim) Run(ctx context.Context, simulated, wait time.Duration) error {
	ms := simulated.Milliseconds()
	if err := s.w.Do(ctx, "sim.run", map[string]any{"for_ms": float64(ms)}); err != nil {
		return err
	}
	return s.WaitStopped(ctx, wait)
}

// Settle steps the engine on a paused run, which is how a command gets the
// time it needs to be answered without starting the clock.
func (s Sim) Settle(ctx context.Context, steps int) error {
	return s.w.Do(ctx, "sim.settle", map[string]any{"steps": float64(steps)})
}

// WaitStopped waits for a run to end.
func (s Sim) WaitStopped(ctx context.Context, timeout time.Duration) error {
	return waitFor(ctx, timeout, "the run to finish", func() (bool, string, error) {
		st, err := s.State(ctx)
		if err != nil {
			return false, "", err
		}
		if !st.Playing {
			return true, "", nil
		}
		return false, secs(st.NowMs) + " of simulated time", nil
	})
}

// WaitUntil waits for the mesh's clock to reach a moment.
func (s Sim) WaitUntil(ctx context.Context, atMs uint32, timeout time.Duration) error {
	return waitFor(ctx, timeout, "simulated time to reach "+secs(atMs),
		func() (bool, string, error) {
			st, err := s.State(ctx)
			if err != nil {
				return false, "", err
			}
			if st.NowMs >= atMs {
				return true, "", nil
			}
			return false, secs(st.NowMs), nil
		})
}

// secs is a simulated moment, said the way a person reads it.
func secs(ms uint32) string {
	return (time.Duration(ms) * time.Millisecond).String()
}

File: pkg/client-js/lib/sim.mjs
// The clock, and the run.

import { MeshbenchError } from "./errors.mjs";
import {
  FIRMWARE_WAIT_MS, JOB_WAIT_MS, RUN_WAIT_MS, secs, waitFor,
} from "./wait.mjs";

/** The clock, and the run. Live. */
export class Sim {
  constructor(wb) { this._wb = wb; }

  /** What the clock is doing. */
  async state() { return (await this._wb.call("sim.state")) || {}; }

  async playing() { return Boolean((await this.state()).playing); }

  async nowMs() { return (await this.state()).now_ms || 0; }

  /** Bring the run up: wait out the warm, start every node, and play.
   *
   *  Deliberately not one call to `sim.start`. That verb is the play button's
   *  own handler and answers four ways - it pauses if already playing, declines
   *  while links are being measured, or starts firmware and does not play - so
   *  a script pressing it once gets whichever of those the moment happens to be
   *  in.
   *
   *  Worse, it only starts firmware when no node is running. Pin a build onto
   *  two nodes of a fifty-eight node fixture and it considers the mesh started,
   *  plays with fifty-six of them down, and says nothing.
   *
   *  So this asks for the three things it actually wants, in order, and checks
   *  each one. */
  async start({ warmMs = JOB_WAIT_MS, firmwareMs = FIRMWARE_WAIT_MS } = {}) {
    // The links first. Nothing that follows means anything against a matrix
    // that is still being measured.
    await this._wb.waitIdle(warmMs);
    // Idle is not the same as measured. A warm that stopped to ask permission
    // to download terrain finishes its own job row, so the wait above returns
    // in a moment having waited for nothing: no link was measured, and every
    // study after this would answer over free space.
    const held = await this.state();
    if (held.warm_held) {
      const note = (held.ground || {}).note ||
        "call terrain.allow to answer the question either way";
      throw new MeshbenchError(
        "the link measurement is held: no terrain has been downloaded and no " +
        `link has been measured. ${note}`);
    }
    // Then every node that is not up, which firmware.start does and sim.start
    // does only when none of them are.
    const st = await this._wb.firmware.state();
    if ((st.running || 0) < (st.nodes || 0)) {
      await this._wb.firmware.start();
      await this._wb.firmware.waitStarted(firmwareMs);
    }
    // Then the clock, by its own name. play cannot pause, which is the other
    // half of what made sim.start unusable from a script.
    if (!(await this.playing())) await this.play();
  }

  async play() { await this._wb.call("sim.play"); }

  async pause() { await this._wb.call("sim.pause"); }

  /** Press the play button as the window presses it, and answer as it answers.
   *
   *  Almost never what a script means - `start()` is - and here because the
   *  verb exists and a client that hid it would send somebody back to `call`. */
  async toggle() { return (await this._wb.call("sim.toggle")) || {}; }

  /** Advance one tick, which is `step_ms` of simulated time. */
  async step() { await this._wb.call("sim.step"); }

  /** Put the clock and the counters back to the start of the run. */
  async reset() { await this._wb.call("sim.reset"); }

  /** Step a paused run, which is how a command gets the time it needs to be
   *  answered without starting the clock. */
  async settle(steps = 60) { await this._wb.call("sim.settle", { steps }); }

  /** Fix the run. Same seed, same scenario, same result - which is what makes a
   *  changed result mean something. */
  async setSeed(seed) { await this._wb.call("sim.seed", { seed }); }

  /** How much simulated time one tick advances. */
  async setStepMs(ms) { await this._wb.call("sim.speed", { step_ms: ms }); }

  /** Whether play starts MeshCore on every node, or runs the channel with
   *  nothing behind it. */
  async setRealFirmware(on = true) { await this._wb.call("sim.kind", { real: on }); }

  /** Advance the mesh's own clock by this much, and wait for it.
   *
   *  Two clocks, one call, and they are not the same one. `simulatedMs` is the
   *  mesh's: five minutes here is five minutes of its time. `waitMs` is yours -
   *  how long you are prepared to sit here before giving up. On 155 emulated
   *  nodes five simulated minutes is a great deal more than five of yours,
   *  which is why the second is separate and generous. */
  async run(simulatedMs, { waitMs = RUN_WAIT_MS } = {}) {
    if (!(simulatedMs > 0)) {
      throw new MeshbenchError("run() needs a length in simulated milliseconds");
    }
    await this._wb.call("sim.run", { for_ms: simulatedMs });
    await this.waitStopped(waitMs);
  }

  /** Wait for a run to end. `timeoutMs` is wall clock. */
  waitStopped(timeoutMs = RUN_WAIT_MS) {
    return waitFor(async () => {
      const st = await this.state();
      if (!st.playing) return [true, ""];
      return [false, `${secs(st.now_ms || 0)} of simulated time`];
    }, timeoutMs, "the run to finish");
  }

  /** Wait for the mesh's clock to reach a moment. `atMs` is simulated time;
   *  `timeoutMs` is yours. */
  waitUntil(atMs, timeoutMs = RUN_WAIT_MS) {
    return waitFor(async () => {
      const st = await this.state();
      if ((st.now_ms || 0) >= atMs) return [true, ""];
      return [false, secs(st.now_ms || 0)];
    }, timeoutMs, `simulated time to reach ${secs(atMs)}`);
  }
}
